### PR TITLE
Add metrics for count chains attestation for issue#1028

### PR DESCRIPTION
**Pull request by sudhishmk**
# Changes -: Work In Progress

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

``` release-note
NONE
```

### CODE REVIEW

**Review by khrm — COMMENTED** [review]
/ok-to-test

File: pkg/reconciler/taskrun/controller.go [hunk 1]
```diff
@@ -45,6 +46,7 @@ func NewController(ctx context.Context, cmw configmap.Watcher) *controller.Impl
 	c := &Reconciler{
 		TaskRunSigner:     tsSigner,
```
Comment (khrm) [inline pkg/reconciler/taskrun/controller.go:47]: Try to put the Metrics recorder inside signer.

File: pkg/reconciler/pipelinerun/controller.go [hunk 2]
```diff
@@ -50,6 +51,7 @@ func NewController(ctx context.Context, cmw configmap.Watcher) *controller.Impl
 		PipelineRunSigner: psSigner,
```
Comment (khrm) [inline pkg/reconciler/pipelinerun/controller.go:51]: Try to put the metrics recorder inside this signer struct.

**Comment by khrm** [comment]
You can run metrics counter here: https://github.com/tektoncd/chains/blob/592410ceddcac777145662c2ad608ccfaaf76ef3/pkg/chains/signing.go#L174

File: pkg/pipelinerunmetrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	"time"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/metrics"
+)
+
+var (
+	atsCount = stats.Float64("attestation_count",
```
Comment (khrm) [inline pkg/pipelinerunmetrics/metrics.go:31]: Let's have two counter: `sign_count` and `payload_stored_count`.

**Comment by khrm** [comment]
Add the above line in the import of following files:
1. pkg/reconciler/pipelinerun/pipelinerun_test.go
2. pkg/reconciler/taskrun/taskrun_test.go

File: pkg/pipelinerunmetrics/metrics.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+		stats.UnitDimensionless)
+
+	plCountView *view.View
+)
+
+// Recorder holds keys for Tekton metrics
+type Recorder struct {
+	mutex           sync.Mutex
+	initialized     bool
+	ReportingPeriod time.Duration
```
Comment (khrm) [inline pkg/pipelinerunmetrics/metrics.go:48]: We don't require this.

File: pkg/taskrunmetrics/metrics.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+		stats.UnitDimensionless)
+
+	plCountView *view.View
+)
+
+// Recorder is used to actually record TaskRun metrics
+type Recorder struct {
+	mutex           sync.Mutex
+	initialized     bool
+	ReportingPeriod time.Duration
```
Comment (khrm) [inline pkg/taskrunmetrics/metrics.go:50]: Same here. It's not needed.

File: pkg/chains/signing.go [hunk 6]
```diff
@@ -44,6 +46,9 @@ type ObjectSigner struct {
 	Backends          map[string]storage.Backend
 	SecretPath        string
 	Pipelineclientset versioned.Interface
+	// Metrics Recorder config
+	PrMetrics *pipelinerunmetrics.Recorder
```
Comment (khrm) [inline pkg/chains/signing.go:50]: We can have an interface here: recorder. 

This interface has two functions: RecordSignedCountMetrics and RecordPayloadCountMetrics.

File: pkg/reconciler/pipelinerun/controller.go (hunk 2 above)
Comment (sudhishmk) [reply pkg/reconciler/pipelinerun/controller.go]: Done

File: pkg/pipelinerunmetrics/metrics.go (hunk 3 above)
Comment (sudhishmk) [reply pkg/pipelinerunmetrics/metrics.go]: Done

File: pkg/pipelinerunmetrics/metrics.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+)
+
+var (
+	sgCount = stats.Float64("signed_total",
+		"Total number of signed pipelineruns",
+		stats.UnitDimensionless)
+
+	sgCountView *view.View
+
+	plCount = stats.Float64("payload_stored_total",
```
Comment (khrm) [inline pkg/pipelinerunmetrics/metrics.go:37]: We can have one more metrics. `Create_Signed_Total`.

File: pkg/pipelinerunmetrics/metrics.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+		"Total number of stored payloads for pipelineruns",
+		stats.UnitDimensionless)
+
+	plCountView *view.View
+)
+
+// Recorder holds keys for Tekton metrics
+type Recorder struct {
+	mutex           sync.Mutex
```
Comment (khrm) [inline pkg/pipelinerunmetrics/metrics.go:46]: We also don't need mutex.

File: pkg/chains/signing.go [hunk 9]
```diff
@@ -135,7 +140,7 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
 		// Extract all the "things" to be signed.
 		// We might have a few of each type (several binaries, or images)
 		objects := signableType.ExtractObjects(ctx, tektonObj)
-
+		count := 0
```
Comment (khrm) [inline pkg/chains/signing.go:143]: We don't need this count.

File: pkg/pipelinerunmetrics/metrics.go (hunk 4 above)
Comment (sudhishmk) [reply pkg/pipelinerunmetrics/metrics.go]: Done

File: pkg/taskrunmetrics/metrics.go (hunk 5 above)
Comment (sudhishmk) [reply pkg/taskrunmetrics/metrics.go]: Done

File: pkg/chains/signing.go (hunk 6 above)
Comment (sudhishmk) [reply pkg/chains/signing.go]: Updated

File: pkg/pipelinerunmetrics/metrics.go (hunk 7 above)
Comment (sudhishmk) [reply pkg/pipelinerunmetrics/metrics.go]: Created  4 metrics as discussed

File: pkg/chains/signing.go (hunk 9 above)
Comment (sudhishmk) [reply pkg/chains/signing.go]: Done

File: pkg/chains/signing.go [hunk 10]
```diff
@@ -222,11 +235,31 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
...
 	return nil
 }
 
+func measureMetrics(metrictype string, mtr MetricsRecorder, ctx context.Context) {
+	logger := logging.FromContext(ctx)
+	logger.Infof("Type of Metrics Recorder ", mtr, metrictype)
+	if mtr != nil {
+		switch mt := metrictype; mt {
+		case "sgcount":
+			mtr.RecordCountMetrics("sgcount", ctx)
```
Comment (khrm) [inline pkg/chains/signing.go:251]: This just adds complexity. Instead of switch and case in two places, Add four different functions.

Comment (khrm) [reply pkg/chains/signing.go:251]: Or you can try to use Generics if possible.

**Comment by khrm** [comment]
Please move this PR from draft now.

File: pkg/pipelinerunmetrics/metrics.go (hunk 4 above)
Comment (khrm) [reply pkg/pipelinerunmetrics/metrics.go]: Please mark all the resolved comments as resolved.

File: pkg/chains/signing.go (hunk 10 above)
Comment (sudhishmk) [reply pkg/chains/signing.go]: Reduced switch statement to 1

**Comment by sudhishmk** [comment]
> Please move this PR from draft now.
Done

File: pkg/taskrunmetrics/metrics.go [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+func (r *Recorder) RecordCountMetrics(MetricType string, ctx context.Context) {
+	logger := logging.FromContext(ctx)
+	logger.Infof("Recording payload count metrics for taskrun ", ctx)
+
+	if !r.initialized {
+		logger.Errorf("ignoring the metrics recording as recorder not initialized ")
+	}
+	switch mt := MetricType; mt {
+	case "sgcount":
```
Comment (khrm) [inline pkg/taskrunmetrics/metrics.go:140]: We can define these constants in pkg/chains and then reuse them here instead of using string.

File: pkg/pipelinerunmetrics/metrics.go [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+	"sync"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/metrics"
+)
+
+var (
+	sgCount = stats.Float64("pipelinerun_signed_total",
```
Comment (khrm) [inline pkg/pipelinerunmetrics/metrics.go:30]: Name this as `pipelinerun_sign_created_total`.

File: pkg/taskrunmetrics/metrics.go [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/metrics"
+)
+
+const anonymous = "anonymous"
+
+var (
+	sgCountView *view.View
+
+	sgCount = stats.Float64("taskrun_signed_total",
```
Comment (khrm) [inline pkg/taskrunmetrics/metrics.go:34]: Name this as `taskrun_sign_created_total`.

File: pkg/taskrunmetrics/metrics.go [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+	"context"
+	"sync"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/metrics"
+)
+
+const anonymous = "anonymous"
```
Comment (khrm) [inline pkg/taskrunmetrics/metrics.go:29]: Why is this here?

File: pkg/taskrunmetrics/metrics.go [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+			r.initialized = false
+			return
+		}
+	})
+
+	return r, errRegistering
+}
+
+func viewRegister() error {
+	r.mutex.Lock()
```
Comment (khrm) [inline pkg/taskrunmetrics/metrics.go:92]: We don't need this in chains.

File: pkg/taskrunmetrics/metrics.go (hunk 11 above)
Comment (sudhishmk) [reply pkg/taskrunmetrics/metrics.go]: Created constants.go and used these constants

File: pkg/pipelinerunmetrics/metrics.go (hunk 12 above)
Comment (sudhishmk) [reply pkg/pipelinerunmetrics/metrics.go]: Done

File: pkg/taskrunmetrics/metrics.go (hunk 13 above)
Comment (sudhishmk) [reply pkg/taskrunmetrics/metrics.go]: Done

File: pkg/taskrunmetrics/metrics.go (hunk 14 above)
Comment (sudhishmk) [reply pkg/taskrunmetrics/metrics.go]: removed it

File: pkg/taskrunmetrics/metrics.go (hunk 15 above)
Comment (sudhishmk) [reply pkg/taskrunmetrics/metrics.go]: Done

File: pkg/chains/constants.go [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package chains
+
+const (
+	SGCount                 = "sgcount"
```
Comment (lcarva) [inline pkg/chains/constants.go:17]: Can we have a more descriptive name for these first 4 constants?

File: pkg/chains/signing.go [hunk 17]
```diff
@@ -174,6 +179,9 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
 			if err != nil {
 				logger.Error(err)
 				continue
+			} else {
```
Comment (lcarva) [inline pkg/chains/signing.go:182]: superfluous `else`

File: pkg/chains/signing.go [hunk 18]
```diff
@@ -174,6 +179,9 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
...
+				logger.Debugf("Recording count for signed messages for %v", obj)
+				measureMetrics("sgcount", o.Recorder, ctx)
```
Comment (lcarva) [inline pkg/chains/signing.go:184]: Use constant

File: pkg/chains/signing.go [hunk 19]
```diff
@@ -189,6 +197,9 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
 				if err := b.StorePayload(ctx, tektonObj, rawPayload, string(signature), storageOpts); err != nil {
 					logger.Error(err)
 					merr = multierror.Append(merr, err)
+				} else {
+					logger.Debugf("Recording count for stored payload for %v", obj)
+					measureMetrics("stcount", o.Recorder, ctx)
```
Comment (lcarva) [inline pkg/chains/signing.go:202]: Use constant

File: pkg/chains/signing.go [hunk 20]
```diff
@@ -204,8 +215,9 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
 					merr = multierror.Append(merr, err)
 				} else {
 					logger.Infof("Uploaded entry to %s with index %d", cfg.Transparency.URL, *entry.LogIndex)
-
 					extraAnnotations[ChainsTransparencyAnnotation] = fmt.Sprintf("%s/api/v1/log/entries?logIndex=%d", cfg.Transparency.URL, *entry.LogIndex)
+					logger.Debugf("Recording count for signed artifacts for %v", obj)
+					measureMetrics("plcount", o.Recorder, ctx)
```
Comment (lcarva) [inline pkg/chains/signing.go:220]: Use constant

File: pkg/chains/signing.go [hunk 21]
```diff
@@ -222,11 +234,22 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
 	// Now mark the TektonObject as signed
 	if err := MarkSigned(ctx, tektonObj, o.Pipelineclientset, extraAnnotations); err != nil {
 		return err
+	} else {
```
Comment (lcarva) [inline pkg/chains/signing.go:237]: superfluous `else`

File: pkg/chains/signing.go [hunk 22]
```diff
@@ -222,11 +234,22 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
...
+		logger.Infof("Recording count for stored payload for %v", tektonObj)
+		measureMetrics("mrcount", o.Recorder, ctx)
```
Comment (lcarva) [inline pkg/chains/signing.go:239]: Use constant

File: pkg/chains/signing.go [hunk 23]
```diff
@@ -174,6 +179,9 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
 			if err != nil {
 				logger.Error(err)
 				continue
+			} else {
+				logger.Debugf("Recording count for signed messages for %v", obj)
```
Comment (lcarva) [inline pkg/chains/signing.go:183]: Are these debug message necessary? If I need to dig through the logs to debug an issue, these will mostly be noise. Maybe they should be logged at the trace level?

File: pkg/chains/signing.go [hunk 24]
```diff
@@ -37,13 +37,19 @@ type Signer interface {
 	Sign(ctx context.Context, obj objects.TektonObject) error
 }
 
+type MetricsRecorder interface {
+	RecordCountMetrics(MetricType string, ctx context.Context)
```
Comment (lcarva) [inline pkg/chains/signing.go:41]: Context should be the first parameter as per go's convention.

File: pkg/chains/signing.go [hunk 25]
```diff
@@ -222,11 +234,22 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
...
 		return err
+	} else {
+		logger.Infof("Recording count for stored payload for %v", tektonObj)
+		measureMetrics("mrcount", o.Recorder, ctx)
 	}
-
 	return nil
 }
 
+func measureMetrics(metrictype string, mtr MetricsRecorder, ctx context.Context) {
```
Comment (lcarva) [inline pkg/chains/signing.go:244]: Context should be the first parameter.

File: pkg/chains/signing.go [hunk 26]
```diff
@@ -222,11 +234,22 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
 	// Now mark the TektonObject as signed
 	if err := MarkSigned(ctx, tektonObj, o.Pipelineclientset, extraAnnotations); err != nil {
 		return err
+	} else {
+		logger.Infof("Recording count for stored payload for %v", tektonObj)
```
Comment (lcarva) [inline pkg/chains/signing.go:238]: Why is this being logged as `Info` while the others are logged as `Debug`?

File: pkg/chains/signing.go [hunk 27]
```diff
@@ -222,11 +234,22 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
...
+		measureMetrics("mrcount", o.Recorder, ctx)
 	}
-
 	return nil
 }
 
+func measureMetrics(metrictype string, mtr MetricsRecorder, ctx context.Context) {
+	logger := logging.FromContext(ctx)
+	logger.Infof("Type of Metrics Recorder ", mtr, metrictype)
```
Comment (lcarva) [inline pkg/chains/signing.go:246]: Let's remove this to avoid noise, or change it to `Trace` level. Also, every call to this function in this file is preceded by a logged message. Let's not duplicate these messages. Remove one.

File: pkg/chains/constants.go [hunk 28]
```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2020 The Tekton Authors
```
Comment (lcarva) [inline pkg/chains/constants.go:2]: New files should probably have the current year's in the copyright header. Please adjust in the other new files as well.

File: pkg/pipelinerunmetrics/injection.go [hunk 29]
```diff
@@ -0,0 +1,50 @@
...
+		logging.FromContext(ctx).Errorf("Failed to create pipelinerun metrics recorder %v", err)
+	}
+	return context.WithValue(ctx, RecorderKey{}, rec)
+}
+
+// Get extracts the pipelinerunmetrics.Recorder from the context.
+func Get(ctx context.Context) *Recorder {
+	untyped := ctx.Value(RecorderKey{})
+	if untyped == nil {
+		logging.FromContext(ctx).Panic("Unable to fetch *pipelinerunmetrics.Recorder from context.")
```
Comment (lcarva) [inline pkg/pipelinerunmetrics/injection.go:47]: Is panicking the best course of action here?

File: pkg/pipelinerunmetrics/metrics.go [hunk 30]
```diff
@@ -0,0 +1,152 @@
...
+	once           sync.Once
+	r              *Recorder
+	errRegistering error
+)
+
+// NewRecorder creates a new metrics recorder instance
+// to log the PipelineRun related metrics
+func NewRecorder(ctx context.Context) (*Recorder, error) {
+	logger := logging.FromContext(ctx)
+	once.Do(func() {
```
Comment (lcarva) [inline pkg/pipelinerunmetrics/metrics.go:74]: Instead of using this, why not just execute this in `init`? You won't have access to the context, but it's probably fine to not log messages here. If `viewRegister` fails, should that be a fatal error? Doing this in `init` probably means you don't need the `initialized` attribute in the `Recorder`.

File: pkg/pipelinerunmetrics/metrics.go [hunk 31]
```diff
@@ -0,0 +1,152 @@
...
+	}
+	return view.Register(
+		sgCountView,
+		plCountView,
+		stCountView,
+		mrCountView,
+	)
+}
+
+func viewUnregister() {
```
Comment (lcarva) [inline pkg/pipelinerunmetrics/metrics.go:121]: I don't see this being used.

File: pkg/pipelinerunmetrics/metrics.go [hunk 32]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+func (r *Recorder) RecordCountMetrics(MetricType string, ctx context.Context) {
+	logger := logging.FromContext(ctx)
+	logger.Debugf("Recording payload count metrics for taskrun ", ctx)
+
+	if !r.initialized {
+		logger.Errorf("ignoring the metrics recording as recorder not initialized ")
+	}
+	switch mt := MetricType; mt {
```
Comment (lcarva) [inline pkg/pipelinerunmetrics/metrics.go:132]: Add a `default` and log an error?

File: pkg/pipelinerunmetrics/metrics.go [hunk 33]
```diff
@@ -0,0 +1,152 @@
...
+	)
+}
+
+func viewUnregister() {
+	view.Unregister(sgCountView, plCountView, stCountView, mrCountView)
+}
+
+func (r *Recorder) RecordCountMetrics(MetricType string, ctx context.Context) {
+	logger := logging.FromContext(ctx)
+	logger.Debugf("Recording payload count metrics for taskrun ", ctx)
```
Comment (lcarva) [inline pkg/pipelinerunmetrics/metrics.go:127]: `Trace` or remove it?

File: pkg/pipelinerunmetrics/metrics.go [hunk 34]
```diff
@@ -0,0 +1,152 @@
...
+		r.countMetrics(ctx, stCount)
+	case chains.MRCount:
+		r.countMetrics(ctx, mrCount)
+	}
+
+}
+
+func (r *Recorder) countMetrics(ctx context.Context, measure *stats.Float64Measure) {
+	logger := logging.FromContext(ctx)
+	logger.Debugf("Recording metrcis for Pipelineruns", measure.Description())
```
Comment (lcarva) [inline pkg/pipelinerunmetrics/metrics.go:147]: This seems unnecessary?

File: pkg/pipelinerunmetrics/metrics.go [hunk 35]
```diff
@@ -0,0 +1,152 @@
...
+	if !r.initialized {
```
Comment (lcarva) [inline pkg/pipelinerunmetrics/metrics.go:148]: Since this is checked in `RecordCountMetrics`, this seems unreachable.

**Review by lcarva — COMMENTED** [review]
Overall, this seems reasonable. My main concern here is the amount of log messages being produced when each metric is recorded.

I left various comments on the pipelinerunmetrics which also apply to the taskrunmetrics.

File: pkg/chains/constants.go (hunk 16 above)
Comment (sudhishmk) [reply pkg/chains/constants.go]: Done

File: pkg/chains/signing.go (hunk 18 above)
Comment (sudhishmk) [reply pkg/chains/signing.go]: Done

File: pkg/chains/signing.go (hunk 20 above)
Comment (sudhishmk) [reply pkg/chains/signing.go]: Done

File: pkg/pipelinerunmetrics/metrics.go (hunk 31 above)
Comment (sudhishmk) [reply pkg/pipelinerunmetrics/metrics.go]: I was testing for any use case which may require unregistering, but didn't find any. So removed it.

File: pkg/pipelinerunmetrics/metrics.go (hunk 32 above)
Comment (sudhishmk) [reply pkg/pipelinerunmetrics/metrics.go:127]: Done

File: pkg/pipelinerunmetrics/metrics.go (hunk 35 above)
Comment (sudhishmk) [reply pkg/pipelinerunmetrics/metrics.go]: Done

File: pkg/chains/signing.go (hunk 17 above)
Comment (sudhishmk) [reply pkg/chains/signing.go]: Done

File: pkg/chains/signing.go (hunk 19 above)
Comment (sudhishmk) [reply pkg/chains/signing.go]: Done

File: pkg/pipelinerunmetrics/metrics.go (hunk 34 above)
Comment (sudhishmk) [reply pkg/pipelinerunmetrics/metrics.go]: removed

**Comment by lcarva** [comment]
@khrm, I see some comments have not yet been addressed. (GitHub collapsed some of them. You'll have to click on "Load more")

File: pkg/chains/signing.go [hunk 36]
```diff
@@ -175,6 +180,8 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
 				logger.Error(err)
 				continue
 			}
+			logger.Debugf("Recording count of signed messages for %v", obj)
```
Comment (lcarva) [inline pkg/chains/signing.go:183]: Trace level?

File: pkg/chains/signing.go [hunk 37]
```diff
@@ -189,6 +196,9 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
 				if err := b.StorePayload(ctx, tektonObj, rawPayload, string(signature), storageOpts); err != nil {
 					logger.Error(err)
 					merr = multierror.Append(merr, err)
+				} else {
+					logger.Debugf("Recording count of stored payload for %v", obj)
```
Comment (lcarva) [inline pkg/chains/signing.go:200]: Trace level?

File: pkg/chains/signing.go [hunk 38]
```diff
@@ -204,8 +214,9 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
 					merr = multierror.Append(merr, err)
 				} else {
 					logger.Infof("Uploaded entry to %s with index %d", cfg.Transparency.URL, *entry.LogIndex)
-
 					extraAnnotations[ChainsTransparencyAnnotation] = fmt.Sprintf("%s/api/v1/log/entries?logIndex=%d", cfg.Transparency.URL, *entry.LogIndex)
+					logger.Debugf("Recording count of uploaded artifacts for %v", obj)
```
Comment (lcarva) [inline pkg/chains/signing.go:218]: Trace level?

File: pkg/chains/signing.go [hunk 39]
```diff
@@ -223,10 +234,20 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
 	if err := MarkSigned(ctx, tektonObj, o.Pipelineclientset, extraAnnotations); err != nil {
 		return err
 	}
-
+	logger.Debugf("Recording count of marked payload for %v", tektonObj)
```
Comment (lcarva) [inline pkg/chains/signing.go:237]: Trace level?

File: pkg/chains/signing.go [hunk 40]
```diff
@@ -223,10 +234,20 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
...
+	measureMetrics(MarkedAsSignedCount, o.Recorder, ctx)
 	return nil
 }
 
+func measureMetrics(metrictype string, mtr MetricsRecorder, ctx context.Context) {
+	logger := logging.FromContext(ctx)
+	logger.Debugf("Type of Metrics Recorder ", mtr, metrictype)
```
Comment (lcarva) [inline pkg/chains/signing.go:244]: I think there's too much logging happening. Can we remove this one?

File: pkg/pipelinerunmetrics/metrics.go [hunk 41]
```diff
@@ -0,0 +1,144 @@
...
+	case chains.SignedMessagesCount:
+		r.countMetrics(ctx, sgCount)
+	case chains.PayloadUploadeCount:
+		r.countMetrics(ctx, plCount)
+	case chains.SignsStoredCount:
+		r.countMetrics(ctx, stCount)
+	case chains.MarkedAsSignedCount:
+		r.countMetrics(ctx, mrCount)
+	default:
+		logger.Errorf("Ignoring the metrics recording as valid Metric type was not found")
```
Comment (lcarva) [inline pkg/pipelinerunmetrics/metrics.go:137]: Let's also add to the message the `mt` value that was used.

File: pkg/pipelinerunmetrics/metrics.go [hunk 42]
```diff
@@ -0,0 +1,144 @@
...
+	}
+	return view.Register(
+		sgCountView,
+		plCountView,
+		stCountView,
+		mrCountView,
+	)
+}
+
+func (r *Recorder) RecordCountMetrics(MetricType string, ctx context.Context) {
```
Comment (lcarva) [inline pkg/pipelinerunmetrics/metrics.go:121]: `MetricType` should probably be something like `metricType`, or just `mt`.

File: pkg/chains/signing.go (hunk 36 above)
Comment (sudhishmk) [reply pkg/chains/signing.go:183]: The reason I used Debug level was because, I didn't find trace level used anywhere in Tekton project. On cross checking the logger library it seems like TRACE level logging is not available for the logger used. Let me know if you want to remove these loggers completely, although I think debug level would good enough as it would help with debugging customer issues faster. Done remaining changes suggested.

File: pkg/chains/signing.go (hunk 40 above)
Comment (sudhishmk) [reply pkg/chains/signing.go]: Removed

File: pkg/pipelinerunmetrics/metrics.go (hunk 42 above)
Comment (sudhishmk) [reply pkg/pipelinerunmetrics/metrics.go]: Done

File: pkg/pipelinerunmetrics/metrics.go (hunk 41 above)
Comment (sudhishmk) [reply pkg/pipelinerunmetrics/metrics.go]: Done. Let me know if I have missed something.

File: pkg/chains/signing.go (hunk 36 above)
Comment (khrm) [reply pkg/chains/signing.go:183]: I think we can remove debug logs from here.

File: pkg/chains/signing.go (hunk 22 above)
Comment (sudhishmk) [reply pkg/chains/signing.go]: Done

File: pkg/chains/signing.go (hunk 24 above)
Comment (sudhishmk) [reply pkg/chains/signing.go]: Done

File: pkg/chains/constants.go (hunk 28 above)
Comment (sudhishmk) [reply pkg/chains/constants.go]: Done

File: pkg/pipelinerunmetrics/metrics.go (hunk 30 above)
Comment (sudhishmk) [reply pkg/pipelinerunmetrics/metrics.go:74]: With init, wouldn't viewRegister  registered at initialization stage by default? in our case we are also calling newrecorder to register client context as well.

File: pkg/chains/signing.go (hunk 36 above)
Comment (sudhishmk) [reply pkg/chains/signing.go]: Removed all debug logging.

File: pkg/chains/signing.go (hunk 37 above)
Comment (sudhishmk) [reply pkg/chains/signing.go]: Removed the loggers as trace level  is not available.

File: pkg/chains/signing.go (hunk 39 above)
Comment (sudhishmk) [reply pkg/chains/signing.go]: Removed the loggers as trace level  is not available.

File: pkg/chains/signing.go (hunk 38 above)
Comment (sudhishmk) [reply pkg/chains/signing.go]: Removed the loggers as trace level  is not available.

File: pkg/pipelinerunmetrics/metrics.go [hunk 43]
```diff
@@ -0,0 +1,144 @@
...
+	initialized bool
+}
+
+// We cannot register the view multiple times, so NewRecorder lazily
+// initializes this singleton and returns the same recorder across any
+// subsequent invocations.
+var (
+	once           sync.Once
+	r              *Recorder
+	errRegistering error
```
Comment (lcarva) [inline pkg/pipelinerunmetrics/metrics.go:67]: Let's scope this variable to the `NewRecorder` function instead of the package. Unless I'm missing something, this doesn't need to be at the package level.

File: pkg/pipelinerunmetrics/metrics.go (hunk 30 above)
Comment (lcarva) [reply pkg/pipelinerunmetrics/metrics.go:74]: The main benefit I see with the current approach is that it allows a context to be passed in which is then used for logging. This wouldn't be possible if the Recorder was added during `init`. Let's leave it as is then.

File: pkg/pipelinerunmetrics/metrics.go (hunk 43 above)
Comment (sudhishmk) [reply pkg/pipelinerunmetrics/metrics.go]: I have moved var "errRegistering" to local scope of func NewRecorder.

File: pkg/pipelinerunmetrics/metrics.go (hunk 30 above)
Comment (sudhishmk) [reply pkg/pipelinerunmetrics/metrics.go:74]: ok

File: pkg/chains/constants.go [hunk 44]
```diff
@@ -0,0 +1,37 @@
...
+*/
+
+package chains
+
+const (
+	SignedMessagesCount     = "sgcount"
+	SignsStoredCount        = "stcount"
+	PayloadUploadeCount     = "plcount"
+	MarkedAsSignedCount     = "mrcount"
+	PipelineRunSignedName   = "pipelinerun_sign_created_total"
```
Comment (khrm) [inline pkg/chains/constants.go:21]: I think the `name` and the `description`  constants for metrics should be in pipelinerun/metrics or taskrun/metrics.

**Review by lcarva — APPROVED** [review]

File: pkg/chains/constants.go (hunk 44 above)
Comment (lcarva) [reply pkg/chains/constants.go:21]: I don't have a strong opinion on this. It might make sense to have them all listed in one place.

**Comment by savitaashture** [comment]
Hi @sudhishmk :wave: 

Is it possible to update metrics doc https://github.com/tektoncd/chains/blob/main/docs/metrics.md so that we will have all the details of the metrics which chains support today 
I reviewed PR yesterday but forgot to add comments

**Comment by sudhishmk** [comment]
> Hi @sudhishmk 👋
> 
> Is it possible to update metrics doc https://github.com/tektoncd/chains/blob/main/docs/metrics.md so that we will have all the details of the metrics which chains support today I reviewed PR yesterday but forgot to add comments

Sure, will do that.
